### PR TITLE
AirfRANS: surface-proximity volume weighting for vol_mse

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
In CFD, near-surface volume predictions (boundary layer, wake region) are more physically important than far-field predictions where the flow is close to freestream. The current volume loss weights all volume points equally, but errors near the airfoil surface matter much more for accurate drag/lift prediction. By weighting the volume loss inversely proportional to distance from the nearest surface point, we focus the model's capacity on the most physically meaningful volume regions.

AF champion (#3135): surface_mse=0.000296, vol_mse=0.002039. Target: vol_mse < 0.0017. Human directive: ALL new AF experiments must target vol_mse < 0.0017.

## Instructions
Champion config baseline: AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, 2L/256d, --model-heads 4, --vol-loss-weight 10.0, --enable-fourier

Implementation:
1. For each volume point, compute the minimum Euclidean distance to any surface point in the same sample.
2. Create a weight: `w_i = 1 / (d_i + epsilon)` where epsilon=0.01. Normalize weights to mean=1.0 to preserve the overall loss scale.
3. Apply: `vol_loss = mean(w_i * (pred_i - target_i)^2)`
4. Alternative form: `w_i = exp(-d_i / scale)` (also normalized to mean=1.0), with scale=0.1 (tight BL focus) or scale=0.5 (broader near-surface region).

Run a smoke test (3 epochs) first to verify shapes and loss values are sensible before running full trials.
Use `--wandb_group af-proximity-volume-weighting` for all runs.

**Trial 1**: Inverse-distance weighting 1/(d+eps), eps=0.01. Full champion config:
```
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 \
  --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0 \
  --wandb_group af-proximity-volume-weighting
```

**Trial 2**: Exponential decay weighting, scale=0.1 (tight boundary layer). Same champion config as Trial 1 but with exp(-d/0.1) weights instead of 1/(d+eps).

**Trial 3** (conditional — only run if T1 or T2 improves over baseline): Take the better-performing weighting scheme from T1/T2 and try scale=0.5 for the exponential form, or eps=0.05 for the inverse-distance form. This tests whether a broader near-surface region performs better than a tighter one.

Hard constraint: surface_mse must not regress beyond 0.000296 — if any trial shows surface regression, stop and report.

## Baseline
- val_primary/surface_mse: 0.000296 (HARD CONSTRAINT — no regressions allowed)
- val_primary/vol_mse: 0.002039 (target: < 0.0017 per human directive)
- Best PR: #3135 (nezuko — EMA=0.999 + vol-weight=10x, 2L/256d, AdamW lr=6e-4, gc=1.0, WD=1e-2)
- W&B baseline run: sh2zyfwr (nezuko/af-ema-vol-weight-10x)
- Reproduce command:
```
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 \
  --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT rei:

## Results

### Trial 1 — Inverse-distance weighting (1/(d + 0.01), clamped)

| Metric | Baseline | T1 | Delta |
|--------|----------|----|-------|
| vol_mse | 0.002039 | 0.0614 | +30.1× ❌ |
| surface_mse | 0.000296 | 0.00457 | +15.4× ❌ |

- **Best epoch:** 88 / 255 trained
- **W&B run:** `aoecdk20`
- **Peak memory:** ~31.7 GB (GPU 0)

### Trial 2 — Exponential decay weighting (exp(-d / 0.1))

| Metric | Baseline | T2 | Delta |
|--------|----------|----|-------|
| vol_mse | 0.002039 | 0.0415 | +20.4× ❌ |
| surface_mse | 0.000296 | 0.000888 | +3.0× ❌ |

- **Best epoch:** 234 / 255 trained
- **W&B run:** `9n27zxp2`
- **Peak memory:** ~38.3 GB (GPU 1)

### Trial 3 — Skipped

Per PR instructions: T3 is conditional on T1 or T2 showing improvement. Neither trial improved over baseline, so T3 was not run.

### Commands used

```bash
# T1 — inverse distance
SENPAI_TIMEOUT_MINUTES=120 CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --ema-decay 0.999 --volume-loss-weight 10.0 \
  --vol-proximity-mode inverse --vol-proximity-eps 0.01 \
  --agent rei --wandb-name "rei/af-prox-T1-inverse-eps001" \
  --wandb-group af-proximity-volume-weighting

# T2 — exponential decay
SENPAI_TIMEOUT_MINUTES=120 CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --ema-decay 0.999 --volume-loss-weight 10.0 \
  --vol-proximity-mode exponential --vol-proximity-scale 0.1 \
  --agent rei --wandb-name "rei/af-prox-T2-exp-scale01" \
  --wandb-group af-proximity-volume-weighting
```

### What happened

**Both trials failed dramatically.** Proximity weighting caused 20–30× regression in vol_mse and 3–15× regression in surface_mse compared to the well-tuned baseline.

**Why it failed:**
1. **Signal imbalance:** The weighting over-concentrates gradient signal on near-surface volume points. Since uniform MSE is the evaluation metric, the model's inability to learn far-field points well dominates the error.
2. **Inverse distance (T1) was worst:** Despite clamping at `median × 20`, the aggressive upweighting near the wall destabilized optimization — best epoch at 88 suggests early convergence to a poor minimum. The 15× surface_mse regression indicates the model's global field quality degraded, not just far-field.
3. **Exponential decay (T2) was more stable** but still far from baseline. Its softer falloff (scale=0.1 means 1/e at ~0.1 distance units) still upweights heavily. Best epoch at 234 shows it continued improving throughout training but couldn't overcome the fundamental signal imbalance.
4. **The current baseline is already well-optimized.** The uniform MSE with `vol_loss_weight=10.0` already provides strong boundary layer learning as part of the full field prediction. Explicit proximity weighting disrupts the natural balance.

### Suggested follow-ups

1. **Warm-start approach:** Train with uniform MSE for most epochs, then switch to proximity weighting only in the final ~10% of training as a fine-tuning step. This preserves the baseline's good global field and adds near-surface refinement.
2. **Much softer weighting:** Try scale=1.0 or 2.0 for exponential, or eps=1.0 for inverse, so the weight ratio between near-surface and far-field is only ~2-3× instead of 10-100×.
3. **Additive auxiliary loss:** Instead of reweighting the main loss, keep uniform MSE as-is and add a small auxiliary loss term computed only on boundary-layer points (e.g., within SDF < 0.05). This avoids disrupting the main optimization.
4. **Curriculum approach:** Gradually increase the proximity weight from 0 to its target value over training, rather than applying full weighting from epoch 0.

### Bug fix included

Fixed a pre-existing `UnboundLocalError` in `main()` where `primary_metric_key` was used as both a function call and a variable name (shadowing). The duplicate variable declarations at lines 1689-1698 were removed and line 1769 was updated to use `best_val_primary_metric_name` instead.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3222 because ADVISOR: Closing — both trials catastrophic. T1 (inverse-distance): surface +647%, vol +2516% worse. T2 (exponential): surface +118%, vol +1766% worse. Root cause: proximity weighting with scale=0.1 or eps=0.01 creates extreme weight ratios that starve far-field gradients (exp(-d/0.1) assigns zero weight to points >0.7 normalized distance) while overdriving near-surface gradients. Since eval uses uniform MSE, the undertrained far-field dominates error. Structural failure, not a hyperparameter issue. Good analysis, rei — the additive auxiliary BL loss suggestion (SDF < 0.05) is the cleaner approach and avoids gradient starvation. Proximity weighting added to AF dead ends..
